Persist state of the channel opening process in FundingManager

halseth opened the pull request:
This PR adds the following persistence to the FundingManager:
- After the funding transaction has been confirmed on-chain, the channel is added to an internal database, marking it `markedOpen`. If the `fundingLocked` message now fails to be sent before a shutdown happen, the FundingManager will retry sending the message on restart.
- After the `fundingLocked` message has been successfully sent, the channel is marked `fundingLockedSent`. Again, if the FundingManager now fails to announce the channel to the network, it will retry upon restart.
- After the announcement has been sent, we delete the channel from this internal database, as the opening process is complete.

Added fundingManager test that goes through the basic, normal workflow without any restarts.

TODO:
- [x] Add test to check restart behavior.
- [x] Update to adhere to `Channel Establishment` in BOLT#2. Note: It specifies that "On disconnection, the funder MUST remember the channel for reconnection if it has broadcast the funding transaction, otherwise it MUST NOT." This PR only adds retransmission of messages after the funding transaction is confirmed.

Addresses #182

Roasbeef commented on fundingmanager_test.go:
```diff
@@ -0,0 +1,596 @@
+	}
+
+	if err != nil {
+		t.Fatalf("erro creating alice fundingmanager: %v", err)
```
erro -> error

Roasbeef commented on fundingmanager_test.go:
```diff
@@ -0,0 +1,596 @@
+	miningNode.TearDown()
+}
+
+func TestFundingManager_normalWorkflow(t *testing.T) {
```
Minor nit: the naming convention for all functions (including test function names), is simply `CamelCase`.

Roasbeef commented on fundingmanager_test.go:
```diff
@@ -0,0 +1,596 @@
+
+	// Create a funding request and start the workflow
+	errChan := make(chan error, 1)
+	updateChan := make(chan *lnrpc.OpenStatusUpdate, 1)
```
At a quick glance, the buffer of this needs to be _two_ otherwise it'll be possible for one of the goroutines to get suck indefinitely on a send.

Roasbeef commented on fundingmanager_test.go:
```diff
@@ -0,0 +1,596 @@
+	}
+
+	// Let Bob handle the init message
+	bobFundingMgr.processFundingRequest(fundingReq.(*lnwire.SingleFundingRequest), aliceAddr)
```
As is now, if the `fundingReq` (and other instances below), _don't_ aren't actually the specified type, then the test will panic. Instead, the interface should be re-assigned to a new variable, with a two-value assignment so the type assertion can be verified: 
```
fundingReqMsg, ok := fundingReq.(*lnwire.SingleFundingRequest)
if !ok {
    t.Fatalf("funding request message not sent, instead %T sent: fundingReq")
}
```

Roasbeef commented on fundingmanager_test.go:
```diff
@@ -0,0 +1,596 @@
+	// forward signature to Alice
+	aliceFundingMgr.processFundingSignComplete(fundinSignComplete.(*lnwire.SingleFundingSignComplete), bobAddr)
+
+	// Small sleep to make sure the transaction is broadcast
```
Rather than using a sleep here, the `ChainNotifier` instance can be used in order to make this test a bit more snappy.

Roasbeef commented on fundingmanager_test.go:
```diff
@@ -0,0 +1,596 @@
+		t.Fatalf("expected fundingLocked sent from Bob")
+	}
+
+	time.Sleep(100 * time.Millisecond)
```
Ideally, we can eliminate this sleep. I think that we can use the `ArbiterChan` to synchronize this exchange instead.

Roasbeef commented on fundingmanager.go:
```diff
@@ -1346,14 +1346,13 @@ func (f *fundingManager) announceChannel(localIDKey, remoteIDKey, localFundingKe
 	// if either fails we consider the announcement unsuccessful.
 	var wg sync.WaitGroup
-	var announcementError error
+	announcementErrors := make(chan error, 3)
```
Ah, nvm my comment then, was looking at an outdated diff 🙃

Roasbeef commented on fundingmanager_test.go:
```diff
@@ -0,0 +1,1046 @@
+	// restart. We mimic this behavior by letting the SendToPeer method return an
+	// error, as if the message was not successfully sent. we then recreate the
+	// fundingManager and make sure it continus the process as expected.
+	aliceFundingMgr.cfg.SendToPeer = func(target *btcec.PublicKey, msgs ...lnwire.Message) error {
```
Looking at the unit test now, the current method is insufficient. The following scenario illustrates the inadequacy of the current logic: we send the funding locked, it fails, now the peer re-connects, in this case we won't re-send the funding locked message until will _restart_. 

To fix this, the `fundingManager` needs access to a callback/subscribe mechanism to be notified once a peer is online or re-connects. With that in place, when the `fundingManager` fails to send the `FundingLocked` messages, then it should request to be notified once the peer comes online once again. If that happens before we restart, then the normal logic should be restarted from that point.

halseth replied:
Discussed offline, and will add such a callback mechanism to the `server`, then do the follow up changes to `fundingManager` in a later PR. Added a `todo` in `fundingManager.Start()`

Roasbeef commented on fundingmanager_test.go:
```diff
@@ -0,0 +1,1046 @@
+	// Run through the process of opening the channel, except sending the
+	// fundingSigned message to Alice. This let us simulate Bob never seeing the
+	// funding transaction being confirmed.
+	errChan := make(chan error, 1)
```
Missing a new line here.

Roasbeef commented on fundingmanager_test.go:
```diff
@@ -0,0 +1,1046 @@
+
+	aliceFundingMgr.initFundingWorkflow(bobAddr, initReq)
+
+	// Alice should have sent the init message to Bob
```
Similar comment here, all comments should be full sentences.

Roasbeef commented on fundingmanager_test.go:
```diff
@@ -0,0 +1,1046 @@
+	}
+
+	// We don't forward this message to Alice, hence the funding transaction is
+	// never broadcasted. Bob wil at this point be waiting for the funding transaction to be
```
wil -> will

halseth replied:
fixed.

halseth commented on fundingmanager.go:
```diff
@@ -229,6 +235,25 @@ type fundingManager struct {
 	wg   sync.WaitGroup
 }
 
+// The channel opening state of a channel represents the different states a
```
done

halseth commented on fundingmanager.go:
```diff
@@ -26,6 +28,10 @@ import (
 	// TODO(roasbeef): tune
 	msgBufferSize = 50
+
+	// Maximum number of blocks to wait for the funding transaction to be
```
done

halseth commented on fundingmanager.go:
```diff
@@ -229,6 +235,25 @@ type fundingManager struct {
+type channelOpeningState uint8
+
+const (
+	// A channel opening state is 'markedOpen' if the funding transaction is
```
done

halseth commented on fundingmanager.go:
```diff
@@ -229,6 +235,25 @@ type fundingManager struct {
+	// other peer.
+	markedOpen channelOpeningState = iota
+
+	// A channel has opening state 'fundingLockedSent' if the fundingLocked
```
done

halseth commented on fundingmanager.go:
```diff
@@ -229,6 +235,25 @@ type fundingManager struct {
+	fundingLockedSent
+)
+
+var channelOpeningStateBucket = []byte("channelOpeningState")
```
fixed

halseth commented on fundingmanager.go:
```diff
@@ -277,7 +302,72 @@ func (f *fundingManager) Start() error {
+
+			select {
+			case <-timeoutChan:
+				// forget the channel, delete it from the database
```
fixed.

halseth commented on fundingmanager.go:
```diff
@@ -277,7 +302,72 @@ func (f *fundingManager) Start() error {
+
+	for _, channel := range openChannels {
+		channelState, shortChanID, err := f.getChannelOpeningState(channel.ChanID)
+		if err == channeldb.ErrChannelNotFound {
```
I thought errors should all be gathered in `errors.go`, but since this one is in package `channeldb` this makes in hindsight not very much sense. Moved the error to `fundingManager.go`.

halseth commented on fundingmanager.go:
```diff
@@ -277,7 +302,72 @@ func (f *fundingManager) Start() error {
+
+		// If we did find the channel in the opening state database, we have seen
+		// the funding transaction being confirmed, but we did not finish the rest
+		// of the setup procedure before we shut down. We handle the remainding
```
arrgh 😡 fixed

halseth commented on fundingmanager.go:
```diff
@@ -794,15 +899,39 @@ func (f *fundingManager) handleFundingComplete(fmsg *fundingCompleteMsg) {
-		f.deleteReservationCtx(peerKey, fmsg.msg.PendingChannelID)
+		select {
+		case <-timeoutChan:
+			// did not see funding confirmation before timeout, forget the channel
```
fixed.

halseth commented on fundingmanager.go:
```diff
@@ -889,13 +1019,64 @@ func (f *fundingManager) handleFundingSignComplete(fmsg *fundingSignCompleteMsg)
+
+			if uint32(epoch.Height) >= maxHeight {
+				fndgLog.Warnf("waited for %v blocks without seeing funding "+
+					"transaction confirmed, cancelling.", maxWaitNumBlocksFundingConf)
```
fixed.

halseth commented on fundingmanager.go:
```diff
@@ -941,6 +1137,38 @@ func (f *fundingManager) waitForFundingConfirmation(completeChan *channeldb.Open
+	// With the block height and the transaction index known, we can
+	// construct the compact chanID which is used on the network to unique
+	// identify channels.
+	shortChanID := lnwire.ShortChannelID{
```
Ok, should I add a TODO here?

Roasbeef replied:
Scratch that. Just confirmed that the current behavior is actually correct. I just pushed a new addition to the interface-level tests (c47047c0b8f50ead8d97f56679674df2b6c8ab3e) to exercise this.

halseth commented on fundingmanager.go:
```diff
@@ -941,6 +1137,38 @@ func (f *fundingManager) waitForFundingConfirmation(completeChan *channeldb.Open
+
+	// The funding transaction now being confirmed, we add this channel to the
+	// fundingManager's internal persistant state machine that we use to track
+	// the remainding process of the channel opening. This is useful to resume
```
🤦‍♂️

halseth commented on fundingmanager.go:
```diff
@@ -941,6 +1137,38 @@ func (f *fundingManager) waitForFundingConfirmation(completeChan *channeldb.Open
+	// fundingManager's internal persistant state machine that we use to track
+	// the remainding process of the channel opening. This is useful to resume
+	// the opening process in case of restarts.
+	err = f.saveChannelOpeningState(completeChan.ChanID, markedOpen, &shortChanID)
```
True true, added TODO. Should I also create a corresponding issue?

halseth commented on fundingmanager.go:
```diff
@@ -1429,3 +1705,89 @@ func copyPubKey(pub *btcec.PublicKey) *btcec.PublicKey {
+
+		bucket := tx.Bucket(channelOpeningStateBucket)
+		if bucket == nil {
+			// If the bucket does not exist, it means we never added a channel to the
```
done

halseth commented on fundingmanager_test.go:
```diff
@@ -0,0 +1,1046 @@
+			return nil, nil
+		},
+	})
+
```
removed.

halseth commented on fundingmanager.go:
```diff
@@ -960,26 +1188,58 @@ func (f *fundingManager) waitForFundingConfirmation(completeChan *channeldb.Open
+
+	// After the channel is successully announced from the fundingManager, we
+	// delete the channel from our internal database. We can do this because we
+	// assume the AnnouncementGossiper queues the announcement messages, and
```
fixed.

halseth commented:
I rebased this on master, and had to add `initLogRotator` to `fundingmanager_test.go` to make it not panic. Unfortunately this leads to some trace logging being produced when running the FundingManager tests.

There were a couple of conflicts related to `localDiscoverySignals` that I tried to resolve after rebasing. In the case we restart and there are channels still in the opening process, I added this before resuming (`fundingmanager.go`, L387):
```go
chanID := lnwire.NewChanIDFromOutPoint(channel.FundingOutpoint)
f.localDiscoverySignals[chanID] = make(chan struct{})
```
Please check that the conflicts were resolved correctly, and that I added those lines of code at the right place (I wasn't sure if the channel barrier was also needed, but I think not, since the funding transaction is already confirmed?)

Roasbeef commented on fundingmanager_test.go:
```diff
@@ -0,0 +1,1021 @@
+	}
+
+	logfile := filepath.Join(logDir, "log")
+	initLogRotator(logfile)
```
If you're disabling the individual loggers, then you don't _also_ need to initialize the log rotator itself.

Roasbeef replied:
I recommend just doing this once at the top of each test.

Roasbeef commented on fundingmanager.go:
```diff
@@ -1190,6 +1199,13 @@ func (f *fundingManager) waitForFundingConfirmation(completeChan *channeldb.Open
+	// remaining part of the opening process, we must wait for this process
+	// to finish (either successully or with some error), before the
+	// fundingManager can be shut down.
+	f.wg.Add(1)
```
This wait group increment should be lifted to the point in which the goroutine is actually launched.

Roasbeef commented on fundingmanager.go:
```diff
@@ -397,14 +388,32 @@ func (f *fundingManager) Start() error {
+			go func() {
+				defer f.wg.Done()
+
+				lnChannel, err := lnwallet.NewLightningChannel(
```
Passing in `ChainNotifier` instance here isn't necessary as this won't be the version of the channel that's used within the `breachArbiter` or `channelLink` structs. Passing in a `nil` parameter for the `ChainNotifier` argument means that a `closeObsever` won't be launched, and no notifications will be requested.
